### PR TITLE
[feature] Add useReset() which returns a function that can clear entire cache

### LINKED ISSUE #145: Empty/Reset the cache on demand
**Is your feature request related to a problem? Please describe.**
Some use cases require us to empty the entire cache e.g. impersonating another user currently it would require calling many invalidates and also setting the invalidIfStale flag.

**Describe the solution you'd like**
For the few instances where this is required I imagine there is a simple way to reset the entire cache with a single hook call.

**Describe alternatives you've considered**
Recreating the entire application context might be possible though it seems heavy handed :)

**Additional context**

### PR DESCRIPTION
**Pull request by blanchg**
Fixes #145 Add Reset Cache hook .

### Motivation

When large amounts of the cache have become invalid it is impossible to clear it all. e.g. the user changes through impersonation.

### Solution

Removing the entire cache means all the resources will request the data immediately and old invalid data won't be shown.

### CODE REVIEW

**Comment by blanchg** [comment]
That last build failed due to a memory error.  Is it a known issue or intermittent issue?

File: src/state/reducer.ts [hunk 1]
```diff
@@ -74,6 +74,13 @@ export default function reducer(
           },
         },
       };
+    case 'rest-hooks/reset': {
```
Comment (ntucker) [inline src/state/reducer.ts]: nit: can you use the initialState constant instead?

File: src/state/__tests__/reducer.ts [hunk 2]
```diff
@@ -261,4 +262,28 @@ describe('reducer', () => {
     const newState = reducer(iniState, action);
     expect(newState).toBe(iniState);
   });
+  it('reset should delete all entities', () => {
```
Comment (ntucker) [inline src/state/__tests__/reducer.ts]: nit: check here is more than just entities so maybe call it 'reset should delete all entries'

File: src/react-integration/hooks/useReset.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { useContext, useCallback } from 'react';
+
+import { DispatchContext } from '~/react-integration/context';
+
+/** Reset the cache */
```
Comment (ntucker) [inline src/react-integration/hooks/useReset.ts]: Returns a function to completely clear the cache of all entries.

File: docs/api/useReset.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
+---
+title: useReset()
+---
+
+<!--DOCUSAURUS_CODE_TABS-->
```
Comment (ntucker) [inline docs/api/useReset.md]: with only one option, no need for code tabs

File: docs/api/useReset.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+// Server returns the logged in user
+export default class CurrentUserResource extends Resource {
+  readonly id: string = null;
+  readonly name: string = '';
+  // ...
+}
+```
+
+```tsx
+
```
Comment (ntucker) [inline docs/api/useReset.md]: nit: extraneous endline

**Comment by ntucker** [comment]
Thanks for the PR, this will be a great contribution! Following the minor notes I will merge this :)

Addressing the memory issue in automated tests.... yeah this started showing up recently. Not sure exactly the nature of it yet. I just reran the tests so it registered it as passed for you.

PS) Can you paste output of yarn test:coverage so we can ensure the tests covered all new lines?

**Comment by blanchg** [comment]
Thanks for the review, adding the changes as requested.

I think adding --maxWorkers=2 to your .circleci/config.yaml:31 will solve your memory problems at the cost of making the testing take a little longer on circleci. https://vgpena.github.io/jest-circleci/

Latest coverage

File                            |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
--------------------------------|----------|----------|----------|----------|-------------------|
All files                       |      100 |    99.48 |     99.4 |      100 |                   |
 src                            |      100 |      100 |      100 |      100 |                   |
  index.ts                      |      100 |      100 |      100 |      100 |                   |
 src/react-integration          |      100 |      100 |      100 |      100 |                   |
  NetworkErrorBoundary.tsx      |      100 |      100 |      100 |      100 |                   |
  context.ts                    |      100 |      100 |      100 |      100 |                   |
  index.ts                      |      100 |      100 |      100 |      100 |                   |
 src/react-integration/hooks    |      100 |    98.46 |      100 |      100 |                   |
  index.ts                      |      100 |      100 |      100 |      100 |                   |
  useCache.ts                   |      100 |      100 |      100 |      100 |                   |
  useCacheNew.ts                |      100 |      100 |      100 |      100 |                   |
  useError.ts                   |      100 |       75 |      100 |      100 |                21 |
  useFetcher.ts                 |      100 |      100 |      100 |      100 |                   |
  useInvalidator.ts             |      100 |      100 |      100 |      100 |                   |
  useMeta.ts                    |      100 |      100 |      100 |      100 |                   |
  useReset.ts                   |      100 |      100 |      100 |      100 |                   |
  useResource.ts                |      100 |      100 |      100 |      100 |                   |
  useResourceNew.ts             |      100 |      100 |      100 |      100 |                   |
  useResultCache.ts             |      100 |      100 |      100 |      100 |                   |
  useRetrieve.ts                |      100 |      100 |      100 |      100 |                   |
  useSubscription.ts            |      100 |      100 |      100 |      100 |                   |
 src/react-integration/provider |      100 |      100 |      100 |      100 |                   |
  CacheProvider.tsx             |      100 |      100 |      100 |      100 |                   |
  ExternalCacheProvider.tsx     |      100 |      100 |      100 |      100 |                   |
  index.ts                      |      100 |      100 |      100 |      100 |                   |
  middleware.ts                 |      100 |      100 |      100 |      100 |                   |
 src/resource                   |      100 |      100 |      100 |      100 |                   |
  Resource.ts                   |      100 |      100 |      100 |      100 |                   |
  SimpleResource.ts             |      100 |      100 |      100 |      100 |                   |
  index.ts                      |      100 |      100 |      100 |      100 |                   |
  normal.js                     |        0 |        0 |        0 |        0 |                   |
  types.ts                      |      100 |      100 |      100 |      100 |                   |
 src/state                      |      100 |      100 |    97.44 |      100 |                   |
  NetworkManager.ts             |      100 |      100 |      100 |      100 |                   |
  PollingSubscription.ts        |      100 |      100 |    88.89 |      100 |                   |
  SubscriptionManager.ts        |      100 |      100 |      100 |      100 |                   |
  reducer.ts                    |      100 |      100 |      100 |      100 |                   |
 src/state/merge                |      100 |      100 |      100 |      100 |                   |
  isMergeable.ts                |      100 |      100 |      100 |      100 |                   |
  mergeDeepCopy.ts              |      100 |      100 |      100 |      100 |                   |
 src/state/selectors            |      100 |      100 |      100 |      100 |                   |
  buildInferredResults.ts       |      100 |      100 |      100 |      100 |                   |
  getEntityPath.ts              |      100 |      100 |      100 |      100 |                   |
  index.ts                      |      100 |      100 |      100 |      100 |                   |
  useDenormalized.ts            |      100 |      100 |      100 |      100 |                   |
  useSchemaSelect.ts            |      100 |      100 |      100 |      100 |                   |

**Comment by ntucker** [comment]
Stable 2.2 release is live https://github.com/coinbase/rest-hooks/releases/tag/2.2.0

**Pull request merged**